### DISCUSSION TITLE
Instantiate LLM model once instead of instantiating for each file

### DISCUSSION THREAD

**Discussion by cmosguy** [jeffmeloy/py2dataset#8] [Ideas]
Hey @jeffmeloy,

So can we modify the lines https://github.com/jeffmeloy/py2dataset/blob/main/py2dataset.py#L83 so that the LLM model is instantiated once outside this function.  The reason being is I am experimenting using a 34B codellama model and it takes about 5 min to load.  If I run this it keeps getting instantiated multiple times.  What do you think?

**Comment by jeffmeloy** [comment]
I've done it this way previously, but when scanning several thousand code files, Python gradually leaks memory somewhere and I found that the simplest approach is to process each file in its own new process. A side effect is that there is some pickling problem that prevents me from passing the model object to the process, so right now the code does it for each file.....it is on my todo to figure out how to fix it, because it bugs me too.

**Comment by jeffmeloy** [comment]
I added a --single_process flag so the code will not create a new process for each code file and only instantiates the model once.  I think good enough for now.

**Comment by cmosguy** [comment]
@jeffmeloy  thanks for doing this - BTW, have you tried doing things not using CTransformers but instead of using `transformers` from HF?  Your code always expects Ctranformers and I am having a ton of problems with that library on our servers.
